cmake/module/Maintenance.cmake

```diff
@@ -30,11 +30,9 @@ function(add_maintenance_targets)
   else()
     set(exe_format ELF)
   endif()
-  if(CMAKE_CROSSCOMPILING)
-    list(JOIN DEPENDS_C_COMPILER_WITH_LAUNCHER " " c_compiler_command)
-  else()
-    set(c_compiler_command ${CMAKE_C_COMPILER})
-  endif()
+  list(JOIN CMAKE_C_COMPILER_LAUNCHER " " c_compiler_command)
```

Comment:
Looks like this has just taken on the depends macos ccache hack behaviour for everything, and I still don't understand why?

Reply:
It is not only about "ccache hack".

Consider the following example:
```
$ cat CMakeLists.txt 
cmake_minimum_required(VERSION 3.22)
set(CMAKE_CXX_COMPILER env -u A_VAR clang++)
project(compiler_demo CXX)
message("CMAKE_CXX_COMPILER: ${CMAKE_CXX_COMPILER}")
$ cmake -B build
-- The CXX compiler identification is Clang 18.1.3
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/env - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMAKE_CXX_COMPILER: /usr/bin/env
-- Configuring done (0.2s)
-- Generating done (0.0s)
-- Build files have been written to: /home/hebasto/CMAKE_COMPILER_DEMO/build
```

The output `CMAKE_CXX_COMPILER: /usr/bin/env` is quite misleading.

Reply:
> set(CMAKE_CXX_COMPILER env -u A_VAR clang++)

Where are we doing this?

Reply:
I still don't understand why we need to do anything other than `set(c_compiler_command ${CMAKE_C_COMPILER})`, like we have been doing this entire time. If this was working perfectly fine before, why does it not work now.

Reply:
> > set(CMAKE_CXX_COMPILER env -u A_VAR clang++)
> 
> Where are we doing this?

https://github.com/hebasto/bitcoin/blob/7973a670915632b75a6aa16f24f98b936865c48f/depends/hosts/darwin.mk#L87-L94

Reply:
> If this was working perfectly fine before, why does it not work now.

It still work perfectly, when a compiler is invoked by CMake.

If we want to pass a compiler invocation string to an external process, we have to re-assemble it from its components.

Reply:
So this is still just for ccache?

Reply:
> No. It is for test-security-check.py and test-symbol-check.py scripts.

Why is it necessary there? I don't see this documented or mentioned anywhere.

Reply:
https://github.com/hebasto/bitcoin/pull/125#discussion_r1580877646:

> It is not only about "ccache hack".
> 
> Consider the following example:
> 
> ```
> $ cat CMakeLists.txt 
> cmake_minimum_required(VERSION 3.22)
> set(CMAKE_CXX_COMPILER env -u A_VAR clang++)
> project(compiler_demo CXX)
> message("CMAKE_CXX_COMPILER: ${CMAKE_CXX_COMPILER}")
> $ cmake -B build
> -- The CXX compiler identification is Clang 18.1.3
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Check for working CXX compiler: /usr/bin/env - skipped
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> CMAKE_CXX_COMPILER: /usr/bin/env
> -- Configuring done (0.2s)
> -- Generating done (0.0s)
> -- Build files have been written to: /home/hebasto/CMAKE_COMPILER_DEMO/build
> ```
> 
> The output `CMAKE_CXX_COMPILER: /usr/bin/env` is quite misleading.

Reply:
If that is meant to be a reply to "Why is it necessary there? I don't see this documented or mentioned anywhere.", then it's not helpful, because ccache shouldnt be used by those scripts, and isn't even available in the environment where they are run.

Reply:
The config output from ccache itself also isn't relevant for those scripts?

Reply:
I don't understand why ccache is brought to this discussion.

The problem is that depends define a compiler invocation string for darwin as a "env ... clang ...". The CMake recognizes the first executable, which is `/usr/bin/env` as a compiler. That is wrong.

Reply:
> The config output from CMake itself also isn't relevant for those scripts?

It demonstrates that the content of the `CMAKE_CXX_COMPILER` variable could be useless to convey it to an external python script that expects a compiler invocation string.

Reply:
Can we get some documentation.

Reply:
> Can we get some documentation.

Sure. That's what I can suggest to consider:

1. The [`CMAKE_<LANG>_COMPILER`](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER.html) variable docs:
> The full path to the compiler for `LANG`.

> If a non-full path value is supplied then CMake will resolve the full path of the compiler.

2. From the source [code](https://github.com/Kitware/CMake/blob/7220b14a2a21406e3f6521343f86318a3f7ce8a6/Modules/CMakeDetermineCompiler.cmake#L119-L120):
```cmake
    # if CMAKE_${lang}_COMPILER is a list, use the first item as
    # CMAKE_${lang}_COMPILER and the rest as CMAKE_${lang}_COMPILER_ARG1
```

3. The [`<LANG>_COMPILER_LAUNCHER`](https://cmake.org/cmake/help/latest/prop_tgt/LANG_COMPILER_LAUNCHER.html) property docs:
> Specify a semicolon-separated list containing a command line for a compiler launching tool. ...
> Some example tools are distcc and ccache.

The `env` command is a similar wrapper tool.

Reply:
I mean a comment in our code.

Reply:
Sure. A comment has been added.